### PR TITLE
Split merging of parser delta and stack environment

### PR DESCRIPTION
**Pull request by kubouch**
# Description

Previously, the environment from the stack was merged into the permanent state withing `merge_delta()`. This caused issues with hooks not having an up-to-date config because it still used the old environment. If you go to the diff and find the hooks tests, you see I was able to delete extra `""` lines which represent pressing "Enter" in the REPL that you needed to press if you wanted the config changes to affect hooks properly.

Now, the stack environment is merged into the permanent state at the very beginning of each REPL line to ensure that whatever that relies on the environment values from the engine state (e.g., the config or cwd) gets the updated value.

# Tests

Make sure you've done the following:

- [x] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.
- [ ] See if it fixes https://github.com/nushell/nushell/issues/5986 (it doesn't)

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

File: crates/nu-command/src/example_test.rs [hunk 1]
```diff
@@ -57,7 +57,10 @@ pub fn test_examples(cmd: impl Command + 'static) {
     };
 
     let cwd = std::env::current_dir().expect("Could not get current working directory.");
-    let _ = engine_state.merge_delta(delta, None, &cwd);
+
+    engine_state
+        .merge_delta(delta)
+        .expect("Error merging delta");
```
Comment (jntrnr) [inline crates/nu-command/src/example_test.rs:63]: all these expects will crash the shell if they happen.

Shouldn't we be a bit more resilient? Maybe just print an error and keep going?

Comment (kubouch) [reply crates/nu-command/src/example_test.rs:63]: @jntrnr These are in a test code. I think it's OK to panic early instead of going with a wrong state when running tests.

**Comment by kubouch** [comment]
Converted to draft to accommodate changes done in https://github.com/nushell/nushell/pull/6040 when it gets merged, but otherwise it's ready.
